src/auth/credential.ts

```diff
@@ -364,116 +263,162 @@ export class RefreshTokenCredential implements Credential {
+      throw new FirebaseAppError(
+        AppErrorCodes.INVALID_CREDENTIAL,
+        'Failed to parse refresh token file: ' + error,
+      );
+    }
   }
-}
-
 
-/**
- * ApplicationDefaultCredential implements the process for loading credentials as
- * described in https://developers.google.com/identity/protocols/application-default-credentials
- */
-export class ApplicationDefaultCredential implements FirebaseCredential {
-  private credential_: Credential;
+  constructor(json: object) {
+    copyAttr(this, json, 'clientId', 'client_id');
+    copyAttr(this, json, 'clientSecret', 'client_secret');
+    copyAttr(this, json, 'refreshToken', 'refresh_token');
+    copyAttr(this, json, 'type', 'type');
 
-  constructor(httpAgent?: Agent) {
-    if (process.env.GOOGLE_APPLICATION_CREDENTIALS) {
-      this.credential_ = credentialFromFile(process.env.GOOGLE_APPLICATION_CREDENTIALS, httpAgent);
-      return;
+    let errorMessage;
+    if (typeof this.clientId !== 'string' || !this.clientId) {
+      errorMessage = 'Refresh token must contain a "client_id" property.';
+    } else if (typeof this.clientSecret !== 'string' || !this.clientSecret) {
+      errorMessage = 'Refresh token must contain a "client_secret" property.';
+    } else if (typeof this.refreshToken !== 'string' || !this.refreshToken) {
+      errorMessage = 'Refresh token must contain a "refresh_token" property.';
+    } else if (typeof this.type !== 'string' || !this.type) {
+      errorMessage = 'Refresh token must contain a "type" property.';
     }
 
-    // It is OK to not have this file. If it is present, it must be valid.
-    if (GCLOUD_CREDENTIAL_PATH) {
-      const refreshToken = RefreshToken.fromPath(GCLOUD_CREDENTIAL_PATH);
-      if (refreshToken) {
-        this.credential_ = new RefreshTokenCredential(refreshToken, httpAgent);
-        return;
-      }
+    if (typeof errorMessage !== 'undefined') {
+      throw new FirebaseAppError(AppErrorCodes.INVALID_CREDENTIAL, errorMessage);
     }
+  }
+}
 
-    this.credential_ = new MetadataServiceCredential(httpAgent);
+export function getApplicationDefault(httpAgent?: Agent): Credential {
+  if (process.env.GOOGLE_APPLICATION_CREDENTIALS) {
+    return credentialFromFile(process.env.GOOGLE_APPLICATION_CREDENTIALS, httpAgent);
   }
 
-  public getAccessToken(): Promise<GoogleOAuthAccessToken> {
-    return this.credential_.getAccessToken();
+  // It is OK to not have this file. If it is present, it must be valid.
+  if (GCLOUD_CREDENTIAL_PATH) {
+    const refreshToken = readCredentialFile(GCLOUD_CREDENTIAL_PATH, true);
+    if (refreshToken) {
+      return new RefreshTokenCredential(refreshToken, httpAgent);
+    }
   }
 
-  public getCertificate(): Certificate | null {
-    return tryGetCertificate(this.credential_);
+  return new ComputeEngineCredential(httpAgent);
+}
+
+function copyAttr(to: {[key: string]: any}, from: {[key: string]: any}, key: string, alt: string) {
```

Comment:
This seems like it should be a utility function. Also add some javadoc for some of these function that may not be trivial to figure out. For example, it was not clear to me what `alt` stands for.

Reply:
Added some jsdoc comments. This was also an pre-existing function.